### DISCUSSION TITLE
prisma.create.post() self-referencing-one-to-many foreign key help

### DISCUSSION THREAD

**Discussion by gordod26** [prisma/prisma#10109] [Q&A] [answered]
Hi,

I need help working through this error. I have a one-to many-self-relationship that dictates all the comments/children a post will have. see schema below.

```typescript
model Post {
  id          Int      @id @default(autoincrement())
  createdAt   DateTime @default(now())
  description String
  postedBy    User?    @relation(fields: [postedById], references: [id])
  postedById  Int?
  votes       Vote[]
  comments    Post[] @relation(name: "comment") 
  parent      Post? @relation(name: "comment", fields: [parentId], references: [id])
  parentId    Int?  @default(0) 
}
```

As you can see, the `parentId` defaults to zero. I always want the root post to have a `parentId` of zero to signify that it is a root post because there is no Post with Id 0 for the foreign key to reference. 

When I added the one-many-self-relationship with default 0 my prisma populated correctly, ignoring foreign-key errors in my db as seen below. 

![2021-11-04_09-52](https://user-images.githubusercontent.com/68024259/140336680-501c515c-1868-4918-8e14-39f26ecaef59.png)

The parent field is null? or undefined? because the `parentId = 0` doesn't exist. That is good because those are root posts. The problem is when I try and make a post mutation using gql I get a foreign key error because there is no post with id 0. See mutation below.

```typescript
async function post(parent, args, context, info) {
  const { userId, prisma } = context;

  if (userId === null) throw new Error("Not Authenticated");
  const newPost = await prisma.post.create({
    data: {
      description: args.description,
      postedBy: { connect: { id: userId } },
      parent: { connect: { id: args.parentId } },
    },
  });
  context.pubsub.publish("NEW_POST", newPost);
  return newPost;
}
```

if I replace `args.parentId = 0` I will get the foreign key error in gql, if `args.parentId = 1` I will not get an error but the post is no longer considered a root post (bad).

can someone please offer me some guidance with this issue? I would like to ignore the foreign key error if possible.

I have looked into the referential actions (seen below) to set a default on post but I don't think my use case is the purpose of referential actions so I am not sure where to look...

```typescript
model Post {
  id             Int     @id @default(autoincrement())
  title          String
  authorUsername String? @default("anonymous")
  author         User?   @relation(fields: [authorUsername], references: [username], onUpdate: SetDefault)
}
```

**Comment by gordod26** [comment, marked as answer]
I was able to solve it. not sure how but This is what I changed...
```prisma
model Post {
  id          Int      @id @default(autoincrement())
  createdAt   DateTime @default(now())
  description String
  postedBy    User?    @relation(fields: [postedById], references: [id])
  postedById  Int?
  votes       Vote[]
  parentId    Int?  
  parent      Post? @relation(name: "ParentComments", fields: [parentId], references: [id])
  comments    Post[] @relation(name: "ParentComments") 
}
```
```typescript
async function post(parent, args, context, info) {
  const { userId, prisma } = context;

  if (userId === null) throw new Error("Not Authenticated");
  const newPost = await prisma.post.create({
    data: {
      description: args.description,
      postedBy: { connect: { id: userId } },
      parent: { connect: { id: args.parentId } },
    },
  });
  context.pubsub.publish("NEW_POST", newPost);
  return newPost;
}
```

```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
